File: wargames_cpp/src/Explosion.cpp
#include "Explosion.hpp"
#include <cmath>

constexpr float Explosion::RING_DELAYS[NUM_RINGS];

Explosion::Explosion(float x, float y, const Color& color)
    : m_x(x)
    , m_y(y)
    , m_color(color)
    , m_age(0.0f)
    , m_duration(2.5f)
{
}

void Explosion::update(float dt) {
    m_age += dt;
}

void Explosion::draw(Renderer* renderer) {
    const float maxRadius = 50.0f;
    const float ringDuration = m_duration - RING_DELAYS[NUM_RINGS - 1];
    
    // Draw each ring
    for (int i = 0; i < NUM_RINGS; i++) {
        float ringAge = m_age - RING_DELAYS[i];
        if (ringAge < 0.0f) continue;
        
        float t = ringAge / ringDuration;
        if (t > 1.0f) continue;
        
        // Expanding radius
        float radius = t * maxRadius;
        
        // Fading alpha
        float alpha = 1.0f - t;
        
        Color ringColor(m_color.r, m_color.g, m_color.b, alpha * m_color.a);
        renderer->drawCircleWithGlow(m_x, m_y, radius, ringColor, 4);
    }
    
    // Central flash (brightest at start)
    if (m_age < 0.5f) {
        float flashAlpha = 1.0f - (m_age / 0.5f);
        float flashRadius = 5.0f + m_age * 10.0f;
        
        Color flashColor(m_color.r, m_color.g, m_color.b, flashAlpha);
        renderer->drawCircleWithGlow(m_x, m_y, flashRadius, flashColor, 5);
    }
}

File: wargames_web/src/Explosion.js
/**
 * Explosion - Represents expanding ring explosion effect
 */

import { lonLatToXY, isInViewport, Colors } from './utils.js';

export class Explosion {
    constructor(lat, lon, startTime) {
        this.lat = lat;
        this.lon = lon;
        this.startTime = startTime;
        this.duration = 2.5; // seconds
        this.finished = false;
        
        // 4 ring waves with staggered timing
        this.rings = [
            { delay: 0.0, maxRadius: 80 },
            { delay: 0.3, maxRadius: 100 },
            { delay: 0.6, maxRadius: 120 },
            { delay: 0.9, maxRadius: 140 }
        ];
    }

    update(currentTime) {
        const elapsed = currentTime - this.startTime;
        
        if (elapsed >= this.duration) {
            this.finished = true;
        }
    }

    render(renderer, currentTime, viewport = null) {
        if (!viewport || isInViewport(this.lon, this.lat, viewport)) {
            const elapsed = currentTime - this.startTime;
            const centerPos = lonLatToXY(
                this.lon,
                this.lat,
                renderer.canvas.width,
                renderer.canvas.height,
                viewport
            );
            
            // Render each ring
            for (const ring of this.rings) {
                const ringElapsed = elapsed - ring.delay;
                if (ringElapsed < 0) continue;
                if (ringElapsed > this.duration - ring.delay) continue;
                
                const ringProgress = ringElapsed / (this.duration - ring.delay);
                const radius = ringProgress * ring.maxRadius;
                const alpha = 1.0 - ringProgress;
                
                // Red explosion color with fade
                const color = [...Colors.RED];
                color[3] = alpha * 0.8;
                
                // Multiple glow layers
                color[3] = alpha * 0.2;
                renderer.drawCircle(centerPos.x, centerPos.y, radius + 4, color, 32);
                
                color[3] = alpha * 0.5;
                renderer.drawCircle(centerPos.x, centerPos.y, radius + 2, color, 32);
                
                color[3] = alpha * 0.8;
                renderer.drawCircle(centerPos.x, centerPos.y, radius, color, 32);
                
                // Central flash for early rings
                if (ringElapsed < 0.3) {
                    const flashAlpha = (1.0 - ringElapsed / 0.3) * 0.6;
                    const flashColor = [1.0, 1.0, 1.0, flashAlpha];
                    renderer.drawCircle(centerPos.x, centerPos.y, 10, flashColor, 16);
                }
            }
        }
    }

    isInView(viewport) {
        if (!viewport) return true;
        return isInViewport(this.lon, this.lat, viewport);
    }
}
